# List of Contents
### 01. Importing Libraries
### 02. Importing Datasets
### 03. Data Wrangling 
### 04. Data Dictionary
### 05. Task 4.4

# 01. Importing Libraries

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Datasets

In [2]:
#path= pathway to project folder
path= r'/Users/alexmendrin/Documents/Career Foundry/Data Immersion/Achievement 4/InstaCart Basket Analysis'

In [3]:
#orders dataframe imported
df_ords= pd.read_csv(os.path.join(path, '02 Data', 'Original Data','orders.csv'))

In [4]:
#products dataframe imported
df_prods=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'))

# 03. Data Wrangling

In [40]:
#Dropping eval_set column from orders.csv
df_ords= df_ords.drop(columns= ['eval_set'])

In [6]:
#rename order_dow
df_ords.rename(columns={'order_dow':'orders_day_of_the_week'}, inplace=True)

In [7]:
#change order_id datatype to string
df_ords['order_id']=df_ords['order_id'].astype('str')

In [8]:
#departments dataframe imported
df_dep= pd.read_csv(os.path.join(path, '02 Data', 'Original Data','departments.csv'))

In [9]:
#transposing df_dep
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [10]:
#rename transposed department dataframe
df_dep_t=df_dep.T

In [11]:
#take the first row of df_dep_t for the header
new_header= df_dep_t.iloc[0]

In [12]:
# take the data under the header row for the new df
df_dep_t_new= df_dep_t[1:]

In [13]:
#set the header row as the df header
df_dep_t_new.columns= new_header

In [14]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# 04. Data Dictionary

In [19]:
#turn department.csv into data dictionary
data_dict= df_dep_t_new.to_dict('index')

In [20]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

# 05. Task 4.4

In [42]:
df_ords

order_id user_id  order_number  orders_day_of_the_week  \
0        2539329       1             1                       2   
1        2398795       1             2                       3   
2         473747       1             3                       3   
3        2254736       1             4                       4   
4         431534       1             5                       4   
...          ...     ...           ...                     ...   
3421078  2266710  206209            10                       5   
3421079  1854736  206209            11                       4   
3421080   626363  206209            12                       1   
3421081  2977660  206209            13                       1   
3421082   272231  206209            14                       6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [31]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

### Q.2

In [16]:
# change user_id datatype to string 
df_ords['user_id']=df_ords['user_id'].astype('str')

In [17]:
#check user_id data type
df_ords['user_id'].dtype

dtype('O')

### Q.3

In [43]:
#renamed order_dow to order_days_of_the_week in line6

### Q.4

In [44]:
#frequency of orders throughout the day to find the busiest hour
df_ords['order_hour_of_day'].value_counts()

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

#### Busiest hour of the day would be 10

### Q.5

In [25]:
df_produce= df_prods[df_prods['department_id']==4]

In [26]:
df_produce

product_id                                      product_name  aisle_id  \
30             31                                White Pearl Onions       123   
42             43                               Organic Clementines       123   
44             45                                 European Cucumber        83   
65             66                         European Style Spring Mix       123   
88             89                    Yogurt Fruit Dip Sliced Apples       123   
...           ...                                               ...       ...   
49582       49578                                Black Garlic Bulbs       123   
49623       49619                                        Opo Squash        83   
49639       49635  Baby Food Blueberry, Parsnip & Buckwheat Stage 2        83   
49661       49657                                 Cabernet Tomatoes        83   
49687       49683                                    Cucumber Kirby        83   

       department_id  prices  
30                 4     7.5  
42                 4    11.5  
44                 4    14.3  
65                 4    11.7  
88                 4    12.6  
...              ...     ...  
49582              4     8.0  
49623              4    12.7  
49639              4    12.5  
49661              4     8.3  
49687              4    13.2  

[1684 rows x 5 columns]

#### Department 4 is the produce department, which is found in my data dictionary on Line20. Above are all the products found within the produce section

### Q.6

In [27]:
#dataframe only including breakfast items
df_breakfast= df_prods[df_prods['department_id']==14]

In [28]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

### Q.7

In [29]:
#dataframe of dinner party items
df_dinner_party= df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

In [30]:
df_dinner_party

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

### Q.8

#### The last dataframe that I created (df_dinner_party) has 7650 rows

### Q.9

In [46]:
df_ords

order_id user_id  order_number  orders_day_of_the_week  \
0        2539329       1             1                       2   
1        2398795       1             2                       3   
2         473747       1             3                       3   
3        2254736       1             4                       4   
4         431534       1             5                       4   
...          ...     ...           ...                     ...   
3421078  2266710  206209            10                       5   
3421079  1854736  206209            11                       4   
3421080   626363  206209            12                       1   
3421081  2977660  206209            13                       1   
3421082   272231  206209            14                       6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [49]:
#all information on user_id of 1
df_user_1= df_ords.loc[df_ords['user_id']=='1']

In [50]:
df_user_1

order_id user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329       1             1                       2                  8   
1   2398795       1             2                       3                  7   
2    473747       1             3                       3                 12   
3   2254736       1             4                       4                  7   
4    431534       1             5                       4                 15   
5   3367565       1             6                       2                  7   
6    550135       1             7                       1                  9   
7   3108588       1             8                       1                 14   
8   2295261       1             9                       1                 16   
9   2550362       1            10                       4                  8   
10  1187899       1            11                       4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

### Q.10

In [51]:
df_user_1.describe()

order_number  orders_day_of_the_week  order_hour_of_day  \
count     11.000000               11.000000          11.000000   
mean       6.000000                2.636364          10.090909   
std        3.316625                1.286291           3.477198   
min        1.000000                1.000000           7.000000   
25%        3.500000                1.500000           7.500000   
50%        6.000000                3.000000           8.000000   
75%        8.500000                4.000000          13.000000   
max       11.000000                4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

### Q.12

In [53]:
#export df_ords to prepared data folder
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

### Q.13

In [55]:
#export df_dep_t_new to prepared data folder
df_dep_t_new.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'departments_wrangled.csv'))